nodejs/aws-infra/experimental/loadBalancer.ts

```diff
@@ -0,0 +1,254 @@
+import * as pulumi from "@pulumi/pulumi";
+
+import * as utils from "../utils";
+
+import * as mod from ".";
+
+export type LoadBalancers = aws.ecs.ServiceArgs["loadBalancers"];
+
+export interface ILoadBalancerProvider {
+    portMappings(containerName: string, name: string, parent: pulumi.Resource): pulumi.Input<aws.ecs.PortMapping[]>;
+    loadBalancers(containerName: string, name: string, parent: pulumi.Resource): LoadBalancers;
+
+    // endpoints(): pulumi.Output<aws.apigateway.x.Endpoint[]>;
+    // defaultEndpoint(): pulumi.Output<aws.apigateway.x.Endpoint>;
+}
+
+export interface PortInfo {
+    /**
+     * The cluster this port will be exposed from.
+     */
+    cluster: mod.Cluster;
+    /**
+     * The incoming port where the service exposes the endpoint.
+     */
+    port: number;
+    /**
+     * The target port on the backing container.  Defaults to the value of [port].
+     */
+    targetPort?: number;
+    /**
+     * Whether the port should be exposed externally.  Defaults to `false`.
+     */
+    external?: boolean;
+    /**
+     * The protocol to use for exposing the service:
+     *  `tcp`: Expose TCP externally and to the container.  Will create a network load balancer.
+     *  `http`: Expose HTTP externally and to the container.  Will create application load balancer.
+     *  `https`: Expose HTTPS externally and HTTP to the container.   Will create application load balancer.
+     *
+     * Defaults to 'tcp' if unspecified.
+     */
+    protocol?: "tcp" | "http" | "https";
+
+    /**
+     * The ARN of the default SSL server certificate. Exactly one certificate is required if the
+     * protocol is [https]. For adding additional SSL certificates, see the
+     * [`aws_lb_listener_certificate`
+     * resource](https://www.terraform.io/docs/providers/aws/r/lb_listener_certificate.html).
+     */
+    certificateArn?: string;
+}
+
+export abstract class LoadBalancerProvider implements ILoadBalancerProvider {
+    public abstract portMappings(
+        containerName: string, name: string, parent: pulumi.Resource): pulumi.Input<aws.ecs.PortMapping[]>;
+    public abstract loadBalancers(
+        containerName: string, name: string, parent: pulumi.Resource): LoadBalancers;
+
+    // public abstract endpoints(): pulumi.Output<aws.apigateway.x.Endpoint[]>;
+    // public abstract defaultEndpoint(): pulumi.Output<aws.apigateway.x.Endpoint>;
+
+    public static fromPortInfo(
+            portInfo: PortInfo,
+            args: aws.elasticloadbalancingv2.LoadBalancerArgs = {}) {
+        return new PortInfoLoadBalancerProvider(portInfo, args);
+    }
+}
+
+export class PortInfoLoadBalancerProvider extends LoadBalancerProvider {
+    portMappings: (containerName: string, name: string, parent: pulumi.Resource) => pulumi.Input<aws.ecs.PortMapping[]>;
+    loadBalancers: (containerName: string, name: string, parent: pulumi.Resource) => LoadBalancers;
+    defaultEndpoint: () => pulumi.Output<aws.apigateway.x.Endpoint>;
```

Comment:
interesting augmentation.  We don't require all ILoadBalancerProviders to expose this (i honestly don't know if they can).  However, this specific impl is able to expose this.  This keeps Task/Service from having any awareness of 'endpoitns'.  but it means if code uses 'fromPortInfo' they get back a provider that easily exposes to them this Endpoint value that they can use if they want for things like hooking up APIGateway proxies.